Source/WebKit/WebProcess/GPU/ShapeDetection/RemoteTextDetectorProxy.h

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#if ENABLE(GPU_PROCESS)
+
+#include "Connection.h"
+#include "ShapeDetectionIdentifier.h"
+#include <WebCore/TextDetectorInterface.h>
+#include <wtf/CompletionHandler.h>
+#include <wtf/Ref.h>
+#include <wtf/RefCounted.h>
+
+namespace IPC {
+class Connection;
+}
+
+namespace WebKit {
+class RemoteRenderingBackendProxy;
+}
+
+namespace WebKit::ShapeDetection {
+
+class RemoteTextDetectorProxy : public WebCore::ShapeDetection::TextDetector {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    static Ref<RemoteTextDetectorProxy> create(RemoteRenderingBackendProxy&, ShapeDetectionIdentifier);
+
+    virtual ~RemoteTextDetectorProxy();
+
+private:
+    RemoteTextDetectorProxy(RemoteRenderingBackendProxy&, ShapeDetectionIdentifier);
+
+    RemoteTextDetectorProxy(const RemoteTextDetectorProxy&) = delete;
+    RemoteTextDetectorProxy(RemoteTextDetectorProxy&&) = delete;
+    RemoteTextDetectorProxy& operator=(const RemoteTextDetectorProxy&) = delete;
+    RemoteTextDetectorProxy& operator=(RemoteTextDetectorProxy&&) = delete;
+
+    ShapeDetectionIdentifier backing() const { return m_backing; }
+
+    void detect(CompletionHandler<void(Vector<WebCore::ShapeDetection::DetectedText>&&)>&&) final;
+
+    ShapeDetectionIdentifier m_backing;
+    RemoteRenderingBackendProxy& m_remoteRenderingBackendProxy;
```

Comment:
Should we use `WeakPtr<>` like in `RemoteDisplayListRecorderProxy` or `RemoteImageBufferProxy`?

There was previously a similar crash where RemoteGPUProxy was destructed before other Proxy classes, leading to use of a reference after its destructor was called